libraries/node-core-library/src/Async.ts

```diff
@@ -73,6 +83,106 @@ export class Async {
     return result;
   }
 
+  /**
+   * Given an input array and a `callback` function, invoke the callback to start a
+   * promise for each element in the array.
+   *
+   * @remarks
+   * This API is similar to the system `Array#forEachAsync`, except that each item can have
+   * a weight that determines how many concurrent operations are allowed. `Array#forEachAsync`
+   * is a special case of this method where weight = 1 for all items.
+   *
+   * The maximum number of concurrent operations can still be throttled using
+   * {@link IAsyncParallelismOptions.concurrency}, however it no longer determines the
+   * maximum number of operations that can be in progress at once. Instead, it determines the
+   * number of concurrency units that can be in progress at once. The weight of each operation
+   * determines how many concurrency units it takes up. For example, if the concurrency is 2
+   * and the first operation has a weight of 2, then only one more operation can be in progress.
+   *
+   * If `callback` throws a synchronous exception, or if it returns a promise that rejects,
+   * then the loop stops immediately.  Any remaining array items will be skipped, and
+   * overall operation will reject with the first error that was encountered.
+   *
+   * @param iterable - the array of inputs for the callback function
+   * @param callback - a function that starts an asynchronous promise for an element
+   *   from the array
+   * @param options - options for customizing the control flow
+   */
+  public static async forEachWeightedAsync<TEntry extends IWeightedIterable>(
```

Comment:
This function should be able to use the inner implementation as `forEachAsync` with just a weight of `1` for each element. You could have these two functions (or a singular function, as I suggested elsewhere) call a shared inner implementation, with each element wrapped in a container object that looks like `{ weight: number, element: TEntry }` defined as `{ weight: element.weight, element }` in the weighted case and `{ weight: 1, element }` in the unweighted case.

Reply:
Updated in [06c7679](https://github.com/microsoft/rushstack/pull/4672/commits/06c7679f01b9f9d6267bd9d580f9c401c62ddbb5)